### PR TITLE
Fortinet FortiManager Scripting Module

### PR DESCRIPTION
**Pull request by Ghilli3**
##### SUMMARY

Adding functionality from playbooks to add/delete/execute CLI or TCL scripts on a FortiManager (FMGR) to run against any number of devices managed by the FMGR.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

fmgr_script.py
fortimanager.py

##### ANSIBLE VERSION

```
ansible 2.4.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/cse/PycharmProjects/ansible/lib/ansible/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
Fully developed in-house by Fortinet based on other modules by Fortinet (fortios_config.py) but for a different product.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

### CODE REVIEW

**Review by gundalow — CHANGES_REQUESTED** [review]
All new network modules since 2.4 need tests. Have a look in `test/unit`. Feel free to join #ansible-network on Freenode IRC for help next week.

**Comment by Ghilli3** [comment]
@gundalow - Thank you, I wasn't aware of the requirement for unit tests, but that makes sense.  Starting to write the unit tests today

**Comment by Ghilli3** [comment]
Hi @gundalow - I've added unit tests for the fortimanager scripting module.

**Review by Ghilli3 — COMMENTED** [review]
Submitting unit tests for the scripting module and ensuring new code passed all the Ansibot checks

**Comment by Ghilli3** [comment]
Hi @gundalow - Please let me know if I need to do anything else to continue moving this forward.  I'm not sure if I have to do anything else for the review.  Thank you for your time and help.

**Comment by Ghilli3** [comment]
Hi All, please let me know if I need to do anything else or if you need anything from me for the pending review.  Thank you.

**Review by Ghilli3 — COMMENTED** [review]
Adding unit tests for the network module for FortiManager scripting per the 2.4 requirement.  Added a few extra commits to trigger the Ansibot scan to ensure all tests passed.

**Comment by Ghilli3** [comment]
Is there any progress or status update on this PR?  Is there anything else needed by me for this PR?  Thank you for your help

File: lib/ansible/modules/network/fortimanager/fmgr_script.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    return response
+
+
+def main():
+    argument_spec = dict(
+        adom=dict(required=False, type="str"),
+        vdom=dict(required=False, type="str"),
+        host=dict(required=True, type="str"),
+        password=dict(fallback=(env_fallback, ["ANSIBLE_NET_PASSWORD"]), no_log=True),
+        username=dict(fallback=(env_fallback, ["ANSIBLE_NET_USERNAME"]), no_log=True),
```
Comment (gundalow) [inline lib/ansible/modules/network/fortimanager/fmgr_script.py]: Please remove `, no_log=True` from the username line

Comment (gundalow) [inline lib/ansible/modules/network/fortimanager/fmgr_script.py]: Missing `username` from `DOCUMENTATION`

File: lib/ansible/modules/network/fortimanager/fmgr_script.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+    url = '/dvmdb/adom/{adom}/script/execute'.format(adom=adom)
+    response = fmg.execute(url, datagram)
+    return response
+
+
+def main():
+    argument_spec = dict(
+        adom=dict(required=False, type="str"),
+        vdom=dict(required=False, type="str"),
```
Comment (gundalow) [inline lib/ansible/modules/network/fortimanager/fmgr_script.py:203]: Missing `vdom` from `DOCUMENTATION:`

File: lib/ansible/modules/network/fortimanager/fmgr_script.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'metadata_version': '1.1'}
+
+DOCUMENTATION = '''
+---
+module: fmgr_script
+version_added: "2.5"
+author: Andrew Welsh
+short_description: Add/Edit/Delete and execute scripts
```
Comment (gundalow) [inline lib/ansible/modules/network/fortimanager/fmgr_script.py:32]: `short_description: Add/Edit/Delete and execute scripts on the FortiManager`

File: lib/ansible/modules/network/fortimanager/fmgr_script.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    host: "{{inventory_hostname}}"
+    username: "{{ username }}"
+    password: "{{ password }}"
+    adom: "root"
+    script_name: "TestScript"
+    script_type: "cli"
+    script_target: "remote_device"
+    script_description: "Create by Ansible"
+    script_content: "get system status"
+- name: EXECUTE SCRIPT
```
Comment (gundalow) [inline lib/ansible/modules/network/fortimanager/fmgr_script.py]: Please have a blank line between tasks, makes it a little more readable.

File: test/units/modules/network/fortimanager/test_fmgr_script.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible.modules.network.fortimanager import fmgr_script
+from units.modules.utils import set_module_args
+from .fortimanager_module import TestFortimanagerModule
```
Comment (gundalow) [inline test/units/modules/network/fortimanager/test_fmgr_script.py]: Imports need to be written so the test is gracefully skipped (`raise SkipTest`) if they are not present, see `test/units/modules/network/f5/test_bigip_monitor_snmp_dca.py` for an example

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
@Ghilli3 Hi, Looking fairly close to getting this merged :)

**Review by Ghilli3 — COMMENTED** [review]
Thank you for looking through the code and finding those issues.  I have fixed all the issues you have requested.

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Merged into `devel` will be released in `2.5.0`

**Pull request merged**